vllm/spec_decode/draft_model_runner.py

```diff
@@ -115,8 +120,228 @@ def update_model_input(
+            print("  attn_metadata:")
+            print("    seq_lens_tensor: {}".format(
+                attn_metadata.seq_lens_tensor))
+            print("    slot_mapping: {}".format(attn_metadata.slot_mapping))
+            print("    block_tables: {}".format(attn_metadata.block_tables))
+
+        return new_model_input
+
+    def _can_use_advance_step(self):
+        if not enable_advance_step:
+            return False
+
+        # TODO: Add support for other attn backends
+        if self.attn_backend.get_name() != "flash-attn":
+            return False
+
+        # TODO: Understand what is needed for LORA support
+        if self.lora_config:
+            return False
+
+        return True
+
+    def _get_model(self, model_input):
+        # TODO: Expand to more backends
+        assert model_input.attn_metadata is not None
+        assert isinstance(model_input.attn_metadata, FlashAttentionMetadata)
+
+        # Currently cuda graph is only supported by the decode phase.
+        is_decode = (model_input.attn_metadata.num_prefills == 0
+                     and model_input.attn_metadata.use_cuda_graph)
+        if is_decode:
+            assert model_input.input_tokens is not None
+            graph_batch_size = model_input.input_tokens.shape[0]
+            model_executable = self.graph_runners[graph_batch_size]
+        else:
+            model_executable = self.model
+
+        return model_executable
+
+    @torch.inference_mode()
+    def _execute_model_with_advance_step(
+            self, model_input: ModelInputForGPUWithSamplingMetadata,
+            kv_caches: List[torch.Tensor],
+            num_steps: int) -> Optional[List[SamplerOutput]]:
+
+        outputs: List[SamplerOutput] = []
+        for step in range(num_steps):
+            model_executable = self._get_model(model_input)
+
+            multi_modal_kwargs = model_input.multi_modal_kwargs or {}
+
+            hidden_states = model_executable(
+                input_ids=model_input.input_tokens,
+                positions=model_input.input_positions,
+                kv_caches=kv_caches,
+                attn_metadata=model_input.attn_metadata,
+                **multi_modal_kwargs,
+            )
+
+            # Compute the logits.
+            logits = self.model.compute_logits(hidden_states,
+                                               model_input.sampling_metadata)
+
+            model_input.sampling_metadata.skip_cpu_samples = True
+
+            # Sample the next token.
+            outputs.append(
+                self.model.sample(
+                    logits=logits,
+                    sampling_metadata=model_input.sampling_metadata,
+                ))
+
+            # Prepare the inputs for the next step.
+            if step != num_steps - 1:
+                if step == 0 and model_input.is_prompt:
+                    model_input = self.update_model_input(
+                        model_input, outputs[-1])
```

Comment:
We should remove `update_model_input`, because draft model runner should not cache the sequence group metadata (the current implementation is just a temporary workaround), so ideally we should directly create decoding inputs by referring to the prefill inputs/outputs, but this may introduce more duplicated logic. One alternative I could think of is we only allow step=1 when it's prefill, and I suppose this is already the case, because we will run prefill for both target and draft model with a single step first for kv-cache (cc @cadedaniel to verify). Thus I could just have

```python 
if model_input.is_prompt and num_steps >1:
    raise ValueError("Prefill with multiple steps is not supported")
```

Reply:
Yeah I was able to remove the temporary logic by simply restricting draft_model_runner to run only with decodes. The only case we get prompt + rest decodes is when we just started the sequence. So by adding this restriction, the draft_model_runner does not run for the first spec step, but for all other steps it does run. I checked the perf penalty, and it has no penalty at all, since most of the time it does run.